### PR TITLE
[WIP: DO NOT MERGE] KAFKA-3522: Allow storing timestamps in RocksDB

### PR DESCRIPTION
**Pull request by mjsax**

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/state/SessionWithTimestampStore.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ */
+public interface SessionWithTimestampStore<K, AGG> extends SessionStore<K, ValueAndTimestamp<AGG>> {
+
+    /**
+     * Write the aggregated value and timestamp for the provided key to the store
+     * @param sessionKey key of the session to write
+     * @param aggregate  the aggregated value for the session, it can be {@code null};
+     *                   if the serialized bytes are also {@code null} it is interpreted as deletes
+     * @param timestamp  the timestamp for the session; ignored if {@code aggregate} is {@code null}
+     * @throws NullPointerException If null is used for sessionKey.
```
Comment (ryannedolan) [inline streams/src/main/java/org/apache/kafka/streams/state/SessionWithTimestampStore.java:34]: I don't think you need to document runtime exceptions like this

File: streams/src/main/java/org/apache/kafka/streams/state/WindowWithTimestampStore.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ * A windowed store interface extending {@link StateStore}.
+ *
+ * In contrast to {@link WindowStore} that stores plain key-value pairs,
+ * {@code WindowWithTimestampStore} stores key-(value/timestamp) pairs.
+ *
+ * @param <K> Type of keys
+ * @param <V> Type of values
+ */
+public interface WindowWithTimestampStore<K, V> extends WindowStore<K, ValueAndTimestamp<V>> {
```
Comment (ryannedolan) [inline streams/src/main/java/org/apache/kafka/streams/state/WindowWithTimestampStore.java:30]: Hard to parse this class name -- sounds like "trait Window with TimestampStore" instead of WindowStore<... ValueAndTimestamp...>

Maybe TimestampWindowStore or TimestampedWindowStore? shrug.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractSegments.java [hunk 3]
```diff
@@ -35,18 +35,19 @@
 import java.util.TreeMap;
 
 /**
- * Manages the {@link Segment}s that are used by the {@link RocksDBSegmentedBytesStore}
+ * Manages the {@link PlainSegment}s or {@link SegmentWithTimestamp} that are used by the {@link RocksDBSegmentedBytesStore}
+ * and {@link RocksDBSegmentedWithTimestampBytesStore}.
  */
-class Segments {
-    private static final Logger log = LoggerFactory.getLogger(Segments.class);
+abstract class AbstractSegments<S extends Segment> {
```
Comment (ryannedolan) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractSegments.java:41]: recommend sticking with T, U, V (or A, B, C for higher-kinded) type parameters

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueWithTimestampStore.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.processor.internals.InternalProcessorContext;
+import org.apache.kafka.streams.processor.internals.ProcessorRecordContext;
+import org.apache.kafka.streams.state.KeyValueStore;
+
+class CachingKeyValueWithTimestampStore<K, V> extends CachingKeyValueStore<K, V> {
```
Comment (ryannedolan) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueWithTimestampStore.java:25]: same here -- sounds like CachingKeyValue with TimestampStore

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueWithTimestampStore.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+    CachingKeyValueWithTimestampStore(final KeyValueStore<Bytes, byte[]> underlying, final Serde<K> keySerde, final Serde<V> valueSerde) {
+        super(underlying, keySerde, valueSerde);
+    }
+
+    void putAndMaybeForward(final ThreadCache.DirtyEntry entry, final InternalProcessorContext context) {
```
Comment (ryannedolan) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueWithTimestampStore.java:31]: large and deeply nested method here -- recommend splitting into multiple smaller private methods where possible

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionWithTimestampStore.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+class CachingSessionWithTimestampStore<K, AGG> extends CachingSessionStore<K, AGG> {
+
+    CachingSessionWithTimestampStore(final SessionStore<Bytes, byte[]> bytesStore,
+                                     final Serde<K> keySerde,
+                                     final Serde<AGG> aggSerde,
+                                     final long segmentInterval) {
+        super(bytesStore, keySerde, aggSerde, segmentInterval);
+    }
+
+    void putAndMaybeForward(final ThreadCache.DirtyEntry entry, final InternalProcessorContext context) {
```
Comment (ryannedolan) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CachingSessionWithTimestampStore.java:37]: same

File: streams/src/main/java/org/apache/kafka/streams/state/internals/ChangeLoggingWindowWithTimestampBytesStore.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.state.WindowStore;
+
+/**
+ * Simple wrapper around a {@link SegmentedBytesStore} to support writing
+ * updates to a changelog
+ */
+class ChangeLoggingWindowWithTimestampBytesStore extends ChangeLoggingWindowBytesStore {
```
Comment (ryannedolan) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/ChangeLoggingWindowWithTimestampBytesStore.java:26]: To avoid these very long class names, consider adding a factory method like ChangeLoggingWindowByteStore.newWithTimestamp() or something

File: streams/src/main/java/org/apache/kafka/streams/state/internals/ChangeLoggingWindowWithTimestampBytesStore.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+    ChangeLoggingWindowWithTimestampBytesStore(final WindowStore<Bytes, byte[]> bytesStore,
```
Comment (ryannedolan) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/ChangeLoggingWindowWithTimestampBytesStore.java:28]: since this wrapper just exposes a new constructor, consider making it a factory method instead

File: streams/src/main/java/org/apache/kafka/streams/state/KeyValueWithTimestampStore.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+ */
+package org.apache.kafka.streams.state;
+
+/**
+ * A key-(value/timestamp) store that supports put/get/delete and range queries.
+ *
+ * @param <K> The key type
+ * @param <V> The value type
+ */
+public interface KeyValueWithTimestampStore<K, V> extends KeyValueStore<K, ValueAndTimestamp<V>> {
```
Comment (ryannedolan) [inline streams/src/main/java/org/apache/kafka/streams/state/KeyValueWithTimestampStore.java:25]: possible alternative: implement `KeyValueStore<K, V>`, and then expose an additional putWithTimestamp, getWithTimestamp etc for callers that want `ValueAndTimestamp<V>` instead of V. This would probably require fewer code changes elsewhere.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueWithTimestampStore.java [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+        // register and possibly restore the state from the logs
+        if (restoreTime.shouldRecord()) {
+            measureLatency(
+                () -> {
+                    inner.init(context, root);
+                    return null;
+                },
+                restoreTime);
+        } else {
+            inner.init(context, root);
```
Comment (ryannedolan) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueWithTimestampStore.java:122]: this pattern of ` if (shouldRecord) measureLatency(X) else X` is not very DRY. You have this same condition (shouldRecord) in several places, and the code for measureLatency(X) vs X is essentially copy-pasted.

Instead, add maybeMeasureLatency(f, sensor) with `if (sensor.shouldRecord()) ... `.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueWithTimestampStore.java [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+                    inner.flush();
+                    return null;
+                },
+                flushTime);
+        } else {
+            inner.flush();
+        }
+    }
+
+    private interface Action<V> {
```
Comment (ryannedolan) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueWithTimestampStore.java:262]: can this just be `Producer<ValueAndTimestamp<V>>` ?

File: streams/src/main/java/org/apache/kafka/streams/state/SessionWithTimestampStore.java (hunk 1 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/SessionWithTimestampStore.java:34]: It common practice in Kafka (at least for Streams API) to document those. Some API accept `null` and others don't. It would be bad user experience to find out via trial-and-error IMHO.

File: streams/src/main/java/org/apache/kafka/streams/state/WindowWithTimestampStore.java (hunk 2 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/WindowWithTimestampStore.java:30]: Thanks for the suggestion. Please follow up on the DISCUSS thread on the dev mailing list with this. All public API changes should be discussed there.
